components/merkledb/src/proof_list_index/proof.rs

```diff
@@ -87,90 +112,52 @@ pub enum ListProofError {
     UnmatchedRootHash,
 }
 
-impl<V> ListProof<V>
+impl<V> ProofVariant<V>
 where
     V: BinaryValue,
 {
     fn collect<'a>(
         &'a self,
         key: ProofListKey,
+        length: u64,
         vec: &mut Vec<(u64, &'a V)>,
     ) -> Result<Hash, ListProofError> {
         if key.height() == 0 {
             return Err(ListProofError::UnexpectedBranch);
         }
         let hash = match *self {
-            ListProof::Full(ref left, ref right) => HashTag::hash_node(
-                &left.collect(key.left(), vec)?,
-                &right.collect(key.right(), vec)?,
+            ProofVariant::Full(ref left, ref right) => HashTag::hash_node(
+                &left.collect(key.left(), length, vec)?,
+                &right.collect(key.right(), length, vec)?,
             ),
-            ListProof::Left(ref left, Some(ref right)) => {
-                HashTag::hash_node(&left.collect(key.left(), vec)?, right)
+            ProofVariant::Left(ref left, Some(ref right)) => {
+                HashTag::hash_node(&left.collect(key.left(), length, vec)?, right)
             }
-            ListProof::Left(ref left, None) => {
-                HashTag::hash_single_node(&left.collect(key.left(), vec)?)
+            ProofVariant::Left(ref left, None) => {
+                HashTag::hash_single_node(&left.collect(key.left(), length, vec)?)
             }
-            ListProof::Right(ref left, ref right) => {
-                HashTag::hash_node(left, &right.collect(key.right(), vec)?)
+            ProofVariant::Right(ref left, ref right) => {
+                HashTag::hash_node(left, &right.collect(key.right(), length, vec)?)
             }
-            ListProof::Leaf(ref value) => {
+            ProofVariant::Leaf(ref value) => {
                 if key.height() > 1 {
                     return Err(ListProofError::UnexpectedLeaf);
                 }
                 vec.push((key.index(), value));
                 HashTag::hash_leaf(&value.to_bytes())
             }
-            ListProof::Absent(ref proof) => {
-                HashTag::hash_list_node(proof.length, proof.merkle_root)
-            }
+            ProofVariant::Absent(merkle_root) => merkle_root,
```

Comment:
This looks incorrect / dangerous at worst, or misleading at best. We should either explicitly check that `Absent` vairant cannot occur anywhere but in the tree root, or rename it to something like `Hash` / `HashedNode` and treat it correspondingly.